### PR TITLE
Add GLSP diagram editor extension

### PR DESCRIPTION
**Pull request by planger**
This extension can generate diagram editors based on the [GLSP examples](https://github.com/eclipse-glsp/glsp-examples) with Java or nodejs backend and package it as Theia extension or VS Code extension.

Co-authored-by: Max Elia <max_elia@hotmail.de>
Co-authored-by: Florian Gareis <dev@zoker.me>

Closes https://github.com/eclipse-theia/generator-theia-extension/issues/110

### CODE REVIEW

File: src/app/index.ts [hunk 1]
```diff
@@ -156,18 +169,58 @@ module.exports = class TheiaExtension extends Base {
...
+                    { value: ExtensionType.DiagramEditor, name: 'DiagramEditor' }
                 ]
             });
             (this.options as any).extensionType = answer.type;
+
+            if (answer.type === ExtensionType.DiagramEditor) {
+                const answer = await this.prompt({
+                    type: 'list',
+                    name: 'backend',
+                    message: 'Which backend do you want?',
```
Comment (JonasHelming) [inline src/app/index.ts:182]:
```suggestion
                    message: 'Which GLSP backend do you want to use, i.e. in which language do you prefer to develop your GLSP server?',
```

File: src/app/index.ts [hunk 2]
```diff
@@ -156,18 +169,58 @@ module.exports = class TheiaExtension extends Base {
...
+                        { value: 'node', name: 'Node' },
+                    ]
+                });
+                let template = answer.backend;
+
+                if(template === 'node') {
+                    const answer = await this.prompt({
+                        type: 'list',
+                        name: 'template',
+                        message: 'Which editor do you want?',
```
Comment (JonasHelming) [inline src/app/index.ts:194]:
```suggestion
                        message: 'Which extension mechanism do you want to use?',
```

File: src/app/index.ts [hunk 3]
```diff
@@ -156,18 +169,58 @@ module.exports = class TheiaExtension extends Base {
...
+                        choices: [
+                            { value: TemplateType.NodeTheia, name: 'Theia' },
```
Comment (JonasHelming) [inline src/app/index.ts:196]:
```suggestion
                            { value: TemplateType.NodeTheia, name: 'Theia extension (more features)' },
```

File: src/app/index.ts [hunk 4]
```diff
@@ -156,18 +169,58 @@ module.exports = class TheiaExtension extends Base {
...
+                            { value: TemplateType.NodeVSCode, name: 'VSCode' },
```
Comment (JonasHelming) [inline src/app/index.ts:197]:
```suggestion
                            { value: TemplateType.NodeVSCode, name: 'VSCode extension (compatible with VS Code)' },
```

File: src/app/index.ts [hunk 5]
```diff
@@ -156,18 +169,58 @@ module.exports = class TheiaExtension extends Base {
...
+                        ]
+                    });
+                    template = answer.template;
+                }
+
+                (this.options as any).templateType = template;
+
+                if(template === TemplateType.Java) {
+                    this.log('\x1b[32m%s\x1b[0m', 'The template will use an EMF source model and can be used for theia ✓')
```
Comment (JonasHelming) [inline src/app/index.ts:206]:
```suggestion
                    this.log('\x1b[32m%s\x1b[0m', 'The template will use an EMF source model on the server and generate a Theia extension ✓')
```

File: src/app/index.ts [hunk 6]
```diff
@@ -156,18 +169,58 @@ module.exports = class TheiaExtension extends Base {
...
+                }
+                if(template === TemplateType.NodeTheia) {
+                    this.log('\x1b[32m%s\x1b[0m', 'The template will use a JSON based source model, node as a server and can be used for Theia ✓')
```
Comment (JonasHelming) [inline src/app/index.ts:209]:
```suggestion
                    this.log('\x1b[32m%s\x1b[0m', 'The template will use a JSON based source model, node as a server and generate a Theia extension ✓')
```

File: src/app/index.ts [hunk 7]
```diff
@@ -156,18 +169,58 @@ module.exports = class TheiaExtension extends Base {
...
+                }
+                if(template === TemplateType.NodeVSCode) {
+                    this.log('\x1b[32m%s\x1b[0m', 'The template will use a JSON based source model, node as a server and can be used for VSCode ✓')
```
Comment (JonasHelming) [inline src/app/index.ts:212]:
```suggestion
                    this.log('\x1b[32m%s\x1b[0m', 'The template will use a JSON based source model, node as a server and generate a VSCode extension ✓')
```

File: src/app/index.ts [hunk 8]
```diff
@@ -156,18 +169,58 @@ module.exports = class TheiaExtension extends Base {
...
             (this.options as any).extensionType = answer.type;
+
+            if (answer.type === ExtensionType.DiagramEditor) {
+                const answer = await this.prompt({
+                    type: 'list',
+                    name: 'backend',
+                    message: 'Which backend do you want?',
+                    choices: [
+                        { value: TemplateType.Java, name: 'Java (requires maven!)' },
+                        { value: 'node', name: 'Node' },
```
Comment (JonasHelming) [inline src/app/index.ts:185]:
```suggestion
                        { value: 'node', name: 'Node (TypeScript)' },
```

**Review by JonasHelming — CHANGES_REQUESTED** [review]
I was unable to generate the GLSP template with Node 12. We should probably introduce a dependecy on >= Node 14

**Comment by planger** [comment]
> I was unable to generate the GLSP template with Node 12. We should probably introduce a dependecy on >= Node 14

Right, I guess that's also true for all other templates as Theia itself requires Node >= 14.

Anyway, we'll add an `engines` entry to the `package.json` of the GLSP project example and move the tag. I'll track that in https://github.com/eclipse-glsp/glsp/issues/676. Thus, this PR doesn't need to be changed in this regard.

**Comment by JonasHelming** [comment]
> > I was unable to generate the GLSP template with Node 12. We should probably introduce a dependecy on >= Node 14
> 
> Right, I guess that's also true for all other templates as Theia itself requires Node >= 14.
> 
> Anyway, we'll add an `engines` entry to the `package.json` of the GLSP project example and move the tag. I'll track that in [eclipse-glsp/glsp#676](https://github.com/eclipse-glsp/glsp/issues/676). Thus, this PR doesn't need to be changed in this regard.

Yes and also: https://github.com/eclipse-theia/generator-theia-extension/issues/154

**Review by JonasHelming — APPROVED** [review]

**Review by vince-fugnitto — COMMENTED** [review]
The following comment brings up similar concerns I had with the previous pull-request (https://github.com/eclipse-theia/generator-theia-extension/pull/111#pullrequestreview-677657569):

> This extension can generate diagram editors based on the [GLSP examples](https://github.com/eclipse-glsp/glsp-examples) with Java or nodejs backend and package it as Theia extension or VS Code extension.

Do we really want to include a generator that is so specific, and which can also provide a vscode extension? The repository is meant to easily get started and create your own Theia extension but now we are starting to mix concepts. For something as specific as `glsp` is it not better in it's own repository? Is there a need to have it in the upstream generator?

**Comment by JonasHelming** [comment]
> The following comment brings up similar concerns I had with the previous pull-request ([#111 (review)](https://github.com/eclipse-theia/generator-theia-extension/pull/111#pullrequestreview-677657569)):
> 
> > This extension can generate diagram editors based on the [GLSP examples](https://github.com/eclipse-glsp/glsp-examples) with Java or nodejs backend and package it as Theia extension or VS Code extension.
> 
> Do we really want to include a generator that is so specific, and which can also provide a vscode extension? The repository is meant to easily get started and create your own Theia extension but now we are starting to mix concepts. For something as specific as `glsp` is it not better in it's own repository? Is there a need to have it in the upstream generator?

For the VS Code extension path, I agree, I would simply remove it. For the Theia extension path, I would very much prefer to have it here. From the view point of a Theia user, it is an interesting capability to explore. Even if the diagram editor is provided by GLSP, it shows a strength of Theia. The generated template is a Theia extension + a Theia example application.
This question we already discussed here: https://github.com/eclipse-theia/generator-theia-extension/pull/111#issuecomment-856635841

So my proposal is to remove the VS Code path and keep the rest as is, are you fine with this? @vince-fugnitto

**Comment by vince-fugnitto** [comment]
@JonasHelming I won't block the pull-request but I'm not entirely confident in the way we consume the `glsp-example` (it would likely be better suited as an example repository) and the fact we need to fetch the `tar.gz` and that it does not include any `template` file in this repository. I'm worried others may do the same and make the repository unmanageable (I already for instance cannot fix bugs in glsp).

If the way forward is to remove the vscode path then let's do that.

**Comment by JonasHelming** [comment]
VS Code template has been removed

**Comment by JonasHelming** [comment]
@vince-fugnitto Would it be possible to do a release of the generator?

**Pull request merged**